client/go/outline/vpn/vpn.go

```diff
@@ -0,0 +1,236 @@
+)
+
+// Config holds the configuration to establish a system-wide [VPNConnection].
+type Config struct {
+	ID              string   `json:"id"`
+	InterfaceName   string   `json:"interfaceName"`
+	IPAddress       string   `json:"ipAddress"`
+	DNSServers      []string `json:"dnsServers"`
+	ConnectionName  string   `json:"connectionName"`
+	RoutingTableId  uint32   `json:"routingTableId"`
+	RoutingPriority uint32   `json:"routingPriority"`
+	ProtectionMark  uint32   `json:"protectionMark"`
+}
+
+// Status defines the possible states of a [VPNConnection].
+type Status string
+
+// Constants representing the different VPN connection statuses.
+const (
+	StatusUnknown       Status = "Unknown"
+	StatusConnected     Status = "Connected"
+	StatusDisconnected  Status = "Disconnected"
+	StatusConnecting    Status = "Connecting"
+	StatusDisconnecting Status = "Disconnecting"
+)
+
+// ProxyDevice is an interface representing a remote proxy server device.
+type ProxyDevice interface {
+	network.IPDevice
+
+	// Connect establishes a connection to the proxy device.
+	Connect(ctx context.Context) error
+
+	// SupportsUDP returns true if the proxy device is able to handle UDP traffic.
+	SupportsUDP() bool
+
+	// RefreshConnectivity refreshes the UDP support of the proxy device.
+	RefreshConnectivity(ctx context.Context) error
+}
+
+// platformVPNConn is an interface representing an OS-specific VPN connection.
+type platformVPNConn interface {
+	// Establish creates a TUN device and routes all system traffic to it.
+	Establish(ctx context.Context) error
+
+	// TUN returns a L3 IP tun device associated with the VPN connection.
+	TUN() io.ReadWriteCloser
+
+	// Close terminates the VPN connection and closes the TUN device.
+	Close() error
+}
+
+// VPNConnection represents a system-wide VPN connection.
+type VPNConnection struct {
+	ID          string `json:"id"`
+	Status      Status `json:"status"`
+	SupportsUDP *bool  `json:"supportsUDP"`
+
+	ctx           context.Context
+	cancel        context.CancelFunc
+	wgEst, wgCopy sync.WaitGroup
+
+	proxy    ProxyDevice
+	platform platformVPNConn
+}
+
+// SetStatus sets the status of the VPN connection.
+func (c *VPNConnection) SetStatus(s Status) {
+	c.Status = s
+}
+
+// SetSupportsUDP sets whether the VPN connection supports UDP.
+func (c *VPNConnection) SetSupportsUDP(v bool) {
+	c.SupportsUDP = &v
+}
+
+// The global singleton VPN connection.
+// This package allows at most one active VPN connection at the same time.
+var mu sync.Mutex
+var conn *VPNConnection
+
+// EstablishVPN establishes a new active [VPNConnection] connecting to a [ProxyDevice]
+// with the given VPN [Config].
+// It first closes any active [VPNConnection] using [CloseVPN], and then marks the
+// newly created [VPNConnection] as the currently active connection.
+// It returns the new [VPNConnection], or an error if the connection fails.
+func EstablishVPN(conf *Config, proxy ProxyDevice) (_ *VPNConnection, err error) {
```

Comment:
Pass the remote device here instead. Or pass the client and create the remote device. You'll need a context for that though.

Reply:
Done.